### PR TITLE
perform initial sync in parallel

### PR DESCRIPTION
**Pull request by stevenctl**
- run in parallel
- log in sync function, we don't do any error handling besides logging anyway

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[X] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by stevenctl** [comment]
> Also concerned about starting 10k goroutines or something

I can do a rate-limit here. 

> I have some concerns about this. The current code is written with the assumption events are processed in serial. I don't know that it will break if we violate this assumption, but we need to be careful.

Maybe an integration test where we deploy echos before istio could sniff out issues.

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 1]
```diff
@@ -466,58 +466,29 @@ func (c *Controller) HasSynced() bool {
...
-		err = multierror.Append(err, c.onServiceEvent(s, model.EventAdd))
-	}
-
-	err = multierror.Append(err, c.syncPods())
-	err = multierror.Append(err, c.syncEndpoints())
-
-	return multierror.Flatten(err.ErrorOrNil())
+func (c *Controller) SyncAll() {
+	syncResources(c.nodeInformer.GetStore(), c.onNodeEvent)
+	syncResources(c.nodeInformer.GetStore(), c.onServiceEvent)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:477]: i think you need service store here

**Comment by hzxuzhonghu** [comment]
As @howardjohn said, we need to control the goroutines

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 1 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: thanks

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -15,12 +15,15 @@
 package controller
 
 import (
+	"context"
 	"fmt"
 	"sort"
 	"sync"
 	"time"
 
+	"github.com/golang/sync/semaphore"
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:24]: import "golang.org/x/sync/semaphore" ?

**Comment by ramaraochavali** [comment]
It may be useful for services and endpoints as we get the benefit parallelism in conversion functions. For node and pod may not be that beneficial as there is not much logic before it is put it in cache - infact for pods since podcache creates a lock it may not benefit at all.

Should we get some numbers and see how much it improves?

**Comment by stevenctl** [comment]
Unless the way I set up this benchmark is flawed, I don't think theres much to gain.

With 100 services, 200 endpoints/200 pods (did not set things up for nodes). 
<pre>
~/g/s/i/i/p/p/s/k/controller ❯❯❯ benchstat sync parallel
name                            old time/op  new time/op   delta
Controller_SyncAll/pod-12       17.2µs ± 2%  102.0µs ± 6%  +493.38%  (p=0.000 n=17+19)
Controller_SyncAll/endpoint-12   274µs ± 2%    165µs ± 7%   -39.80%  (p=0.000 n=18+20)
<strike>Controller_SyncAll/node-12      25.2ns ± 0%  136.0ns ± 1%  +439.33%  (p=0.000 n=18+17)</strike>
Controller_SyncAll/service-12   57.7µs ± 1%   84.5µs ±10%   +46.36%  (p=0.000 n=19+20)
</pre>

**Pull request closed without merging**